Transaction too large

```
03-10 00:11:02.958 D/main    (27424): pausing MainActivity
03-10 00:11:03.096 E/JavaBinder(27424): !!! FAILED BINDER TRANSACTION !!!  (parcel size = 879808)
03-10 00:11:03.097 W/ActivityStopInfo(27424): Bundle stats:
03-10 00:11:03.097 W/ActivityStopInfo(27424):   android:viewHierarchyState [size=1632]
03-10 00:11:03.097 W/ActivityStopInfo(27424):     android:views [size=1584]
03-10 00:11:03.112 W/ActivityStopInfo(27424):   androidx.lifecycle.BundlableSavedStateRegistry.key [size=877236]
03-10 00:11:03.112 W/ActivityStopInfo(27424):     android:support:activity-result [size=2908]
03-10 00:11:03.112 W/ActivityStopInfo(27424):       KEY_COMPONENT_ACTIVITY_REGISTERED_KEYS [size=2192]
03-10 00:11:03.128 W/ActivityStopInfo(27424):     android:support:fragments [size=873952]
03-10 00:11:03.128 W/ActivityStopInfo(27424):       fragment_6194f1ca-5b9f-4e87-8e2e-5e5cd568cce1 [size=2520]
03-10 00:11:03.128 W/ActivityStopInfo(27424):         state [size=2488]
03-10 00:11:03.128 W/ActivityStopInfo(27424):       state [size=1396]
03-10 00:11:03.140 W/ActivityStopInfo(27424):       fragment_d976c0ad-b321-4fcf-a33b-01c4d8779942 [size=796736]
03-10 00:11:03.152 W/ActivityStopInfo(27424):         state [size=796704]
03-10 00:11:03.153 W/ActivityStopInfo(27424):       fragment_9d5b70d0-9955-4fdb-b670-1b50f4dd285c [size=72156]
03-10 00:11:03.154 W/ActivityStopInfo(27424):         state [size=72124]
03-10 00:11:03.154 W/ActivityStopInfo(27424): PersistableBundle stats:
03-10 00:11:03.154 W/ActivityStopInfo(27424):   [null]
03-10 00:11:03.154 D/AndroidRuntime(27424): Shutting down VM
--------- beginning of crash
03-10 00:11:03.155 E/AndroidRuntime(27424): FATAL EXCEPTION: main
03-10 00:11:03.155 E/AndroidRuntime(27424): Process: space.taran.arknavigator, PID: 27424
03-10 00:11:03.155 E/AndroidRuntime(27424): java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 879808 bytes
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.app.ActivityClient.activityStopped(ActivityClient.java:87)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:143)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.Handler.handleCallback(Handler.java:938)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.Handler.dispatchMessage(Handler.java:99)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.Looper.loopOnce(Looper.java:201)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.Looper.loop(Looper.java:288)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.app.ActivityThread.main(ActivityThread.java:7870)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at java.lang.reflect.Method.invoke(Native Method)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
03-10 00:11:03.155 E/AndroidRuntime(27424): Caused by: android.os.TransactionTooLargeException: data parcel size 879808 bytes
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.BinderProxy.transactNative(Native Method)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.os.BinderProxy.transact(BinderProxy.java:571)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.app.IActivityClientController$Stub$Proxy.activityStopped(IActivityClientController.java:1315)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	at android.app.ActivityClient.activityStopped(ActivityClient.java:84)
03-10 00:11:03.155 E/AndroidRuntime(27424): 	... 9 more
```

--- a/app/src/main/java/space/taran/arknavigator/ui/adapter/TagsSelectorAdapter.kt
+++ b/app/src/main/java/space/taran/arknavigator/ui/adapter/TagsSelectorAdapter.kt
@@ -1,14 +1,21 @@
 package space.taran.arknavigator.ui.adapter
 
+import android.content.Context
 import android.content.res.ColorStateList
 import android.graphics.Color
+import android.view.LayoutInflater
 import android.view.View
+import android.view.ViewGroup
 import androidx.core.content.ContextCompat
 import androidx.core.view.isVisible
 import androidx.fragment.app.Fragment
-import com.google.android.material.chip.Chip
+import com.google.android.flexbox.FlexboxLayoutManager
+import com.mikepenz.fastadapter.FastAdapter
+import com.mikepenz.fastadapter.adapters.ItemAdapter
+import com.mikepenz.fastadapter.binding.AbstractBindingItem
 import space.taran.arknavigator.R
 import space.taran.arknavigator.databinding.FragmentResourcesBinding
+import space.taran.arknavigator.databinding.ItemTagBinding
 import space.taran.arknavigator.databinding.PopupResourcesTagMenuBinding
 import space.taran.arknavigator.mvp.presenter.adapter.tagsselector.TagItem
 import space.taran.arknavigator.mvp.presenter.adapter.tagsselector.TagsSelectorPresenter
@@ -24,15 +31,27 @@ class TagsSelectorAdapter(
     @Inject
     lateinit var stringProvider: StringProvider
 
-    private val checkedChipGroup = binding.cgTagsChecked
-    private val chipGroup = binding.tagsCg
+    private val tagsAdapter = ItemAdapter<TagItemView>()
+    private val filterTagsAdapter = ItemAdapter<TagItemView>()
+
     private val clearChip = binding.btnClear
-    private val chipsByTagItems = mutableMapOf<TagItem, Chip>()
+
+    init {
+        binding.rvTags.apply {
+            layoutManager = FlexboxLayoutManager(fragment.requireContext())
+            adapter = FastAdapter.with(tagsAdapter)
+            itemAnimator = null
+        }
+
+        binding.rvTagsFilter.apply {
+            layoutManager = FlexboxLayoutManager(fragment.requireContext())
+            adapter = FastAdapter.with(filterTagsAdapter)
+            itemAnimator = null
+        }
+    }
 
     fun drawTags() {
         drawClearChip()
-        chipGroup.removeAllViews()
-        checkedChipGroup.removeAllViews()
 
         if (checkTagsEmpty()) {
             binding.tvTagsSelectorHint.isVisible = true
@@ -40,10 +59,40 @@ class TagsSelectorAdapter(
         } else
             binding.tvTagsSelectorHint.isVisible = false
 
-        createChips()
-        drawIncludedAndExcludedTags()
-        drawAvailableTags()
-        drawUnavailableTags()
+        val includedTagItemView = presenter.includedTagItems.map {
+            createTagItemView(it, TagSelectType.Included)
+        }
+        val excludedTagItemView = presenter.excludedTagItems.map {
+            createTagItemView(it, TagSelectType.Excluded)
+        }
+        val availableTagItemView = presenter.availableTagsForDisplay.map {
+            createTagItemView(it, TagSelectType.Available)
+        }
+        val unavailableTagItemView = presenter.unavailableTagsForDisplay.map {
+            createTagItemView(it, TagSelectType.Unavailable)
+        }
+
+        val includedAndExcluded = includedTagItemView + excludedTagItemView
+        val includedAndExcludedSorted =
+            presenter.includedAndExcludedTagsForDisplay.map { tagItem ->
+                includedAndExcluded.find { it.tagItem == tagItem }
+                    ?: error("TagsSelectorPresenter: Tag inconsistency detected")
+            }
+
+        if (presenter.filterEnabled) {
+            filterTagsAdapter.set(includedAndExcludedSorted)
+            tagsAdapter.set(
+                availableTagItemView +
+                    unavailableTagItemView
+            )
+        } else {
+            filterTagsAdapter.set(emptyList())
+            tagsAdapter.set(
+                includedAndExcludedSorted +
+                    availableTagItemView +
+                    unavailableTagItemView
+            )
+        }
     }
 
     private fun checkTagsEmpty(): Boolean = with(presenter) {
@@ -51,55 +100,6 @@ class TagsSelectorAdapter(
             excludedTagItems.isEmpty() &&
             availableTagsForDisplay.isEmpty() &&
             unavailableTagsForDisplay.isEmpty()
-    }
-
-    private fun createChips() {
-        chipsByTagItems.clear()
-        presenter.includedTagItems.forEach { tag ->
-            val chip = createDefaultChip(tag)
-            chip.setTextColor(Color.BLUE)
-            chip.isChecked = true
-            chipsByTagItems[tag] = chip
-        }
-        presenter.excludedTagItems.forEach { tag ->
-            val chip = createDefaultChip(tag)
-            chip.setTextColor(Color.RED)
-            chip.isLongClickable = false
-            chipsByTagItems[tag] = chip
-        }
-        presenter.availableTagsForDisplay.forEach { tag ->
-            val chip = createDefaultChip(tag)
-            chipsByTagItems[tag] = chip
-        }
-        presenter.unavailableTagsForDisplay.forEach { tag ->
-            val chip = createDefaultChip(tag)
-            chip.setTextColor(Color.GRAY)
-            chip.isLongClickable = false
-            chip.isClickable = false
-            chip.isCheckable = false
-            chipsByTagItems[tag] = chip
-        }
-    }
-
-    private fun drawIncludedAndExcludedTags() {
-        presenter.includedAndExcludedTagsForDisplay.forEach { tag ->
-            if (presenter.filterEnabled)
-                checkedChipGroup.addView(chipsByTagItems[tag])
-            else
-                chipGroup.addView(chipsByTagItems[tag])
-        }
-    }
-
-    private fun drawAvailableTags() {
-        presenter.availableTagsForDisplay.forEach { tag ->
-            chipGroup.addView(chipsByTagItems[tag])
-        }
-    }
-
-    private fun drawUnavailableTags() {
-        presenter.unavailableTagsForDisplay.forEach { tag ->
-            chipGroup.addView(chipsByTagItems[tag])
-        }
     }
 
     private fun drawClearChip() = clearChip.apply {
@@ -116,37 +116,95 @@ class TagsSelectorAdapter(
         }
     }
 
-    private fun createDefaultChip(item: TagItem) = Chip(chipGroup.context).apply {
+    private fun createTagItemView(tagItem: TagItem, tagType: TagSelectType) =
+        TagItemView(
+            fragment,
+            fragment.requireContext(),
+            presenter,
+            stringProvider,
+            tagItem,
+            tagType
+        )
+}
+
+private enum class TagSelectType {
+    Included, Excluded, Available, Unavailable
+}
+
+private class TagItemView(
+    private val fragment: Fragment,
+    private val ctx: Context,
+    private val presenter: TagsSelectorPresenter,
+    private val stringProvider: StringProvider,
+    val tagItem: TagItem,
+    private val tagType: TagSelectType
+) : AbstractBindingItem<ItemTagBinding>() {
+
+    override val type = R.id.fastadapter_item
+
+    override fun createBinding(
+        inflater: LayoutInflater,
+        parent: ViewGroup?
+    ) = ItemTagBinding.inflate(inflater, parent, false)
+
+    override fun bindView(
+        binding: ItemTagBinding,
+        payloads: List<Any>
+    ) = with(binding.chipTag) {
+
+        resetTagView(binding)
+
+        setOnClickListener {
+            presenter.onTagItemClick(tagItem)
+        }
+
+        setOnLongClickListener {
+            showTagMenuPopup(tagItem, it)
+            true
+        }
+
+        when (tagItem) {
+            is TagItem.PlainTagItem -> {
+                chipBackgroundColor =
+                    ColorStateList.valueOf(
+                        ContextCompat.getColor(ctx, R.color.grayTransparent)
+                    )
+                text = tagItem.tag
+            }
+            is TagItem.KindTagItem -> {
+                chipBackgroundColor =
+                    ColorStateList.valueOf(
+                        ContextCompat.getColor(ctx, R.color.blue)
+                    )
+                text = stringProvider.kindToString(tagItem.kind)
+            }
+        }
+
+        when (tagType) {
+            TagSelectType.Included -> {
+                setTextColor(Color.BLUE)
+                isChecked = true
+            }
+            TagSelectType.Excluded -> {
+                setTextColor(Color.RED)
+                isLongClickable = false
+            }
+            TagSelectType.Available -> {}
+            TagSelectType.Unavailable -> {
+                setTextColor(Color.GRAY)
+                isLongClickable = false
+                isClickable = false
+                isCheckable = false
+            }
+        }
+    }
+
+    private fun resetTagView(binding: ItemTagBinding) = binding.chipTag.apply {
         isClickable = true
         isLongClickable = true
         isCheckable = true
         isChecked = false
         setTextColor(Color.BLACK)
-        when (item) {
-            is TagItem.PlainTagItem -> {
-                chipBackgroundColor =
-                    ColorStateList.valueOf(
-                        ContextCompat.getColor(context, R.color.grayTransparent)
-                    )
-                text = item.tag
-            }
-            is TagItem.KindTagItem -> {
-                chipBackgroundColor =
-                    ColorStateList.valueOf(
-                        ContextCompat.getColor(context, R.color.blue)
-                    )
-                text = stringProvider.kindToString(item.kind)
-            }
-        }
-
-        setOnClickListener {
-            presenter.onTagItemClick(item)
-        }
-
-        setOnLongClickListener {
-            showTagMenuPopup(item, it)
-            true
-        }
     }
 
     private fun showTagMenuPopup(tag: TagItem, tagView: View) {

--- a/app/src/main/java/space/taran/arknavigator/ui/fragments/ResourcesFragment.kt
+++ b/app/src/main/java/space/taran/arknavigator/ui/fragments/ResourcesFragment.kt
@@ -215,7 +215,7 @@ class ResourcesFragment : MvpAppCompatFragment(), ResourcesView {
 
     override fun setTagsFilterEnabled(enabled: Boolean) {
         binding.layoutInput.isVisible = enabled
-        binding.cgTagsChecked.isVisible = enabled
+        binding.rvTagsFilter.isVisible = enabled
         if (enabled) {
             binding.etTagsFilter.placeCursorToEnd()
             binding.etTagsFilter.showKeyboard()

--- a/app/src/main/res/layout/fragment_resources.xml
+++ b/app/src/main/res/layout/fragment_resources.xml
@@ -131,8 +131,8 @@
                 android:layout_height="wrap_content"
                 android:orientation="vertical">
 
-                <com.google.android.material.chip.ChipGroup
-                    android:id="@+id/cg_tags_checked"
+                <androidx.recyclerview.widget.RecyclerView
+                    android:id="@+id/rv_tags_filter"
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
                     android:padding="10dp"
@@ -145,6 +145,7 @@
                     android:layout_height="wrap_content"
                     android:layout_marginStart="10dp"
                     android:layout_marginEnd="10dp"
+                    android:layout_marginBottom="8dp"
                     android:hint="@string/resources_filter_hint"
                     android:visibility="gone"
                     app:startIconDrawable="@drawable/ic_baseline_search_24">
@@ -157,11 +158,11 @@
 
                 </com.google.android.material.textfield.TextInputLayout>
 
-                <com.google.android.material.chip.ChipGroup
-                    android:id="@+id/tags_cg"
+                <androidx.recyclerview.widget.RecyclerView
+                    android:id="@+id/rv_tags"
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
-                    android:padding="10dp"
+                    android:paddingHorizontal="10dp"
                     android:visibility="visible" />
             </LinearLayout>
 

--- a/app/src/main/res/layout/item_tag.xml
+++ b/app/src/main/res/layout/item_tag.xml
@@ -1,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<com.google.android.material.chip.Chip xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/chip_tag"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    android:layout_marginHorizontal="4dp"
+    android:textColor="@color/black"/>

